### ISSUE TITLE
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST value impacts coco_evaluation metrics

### ISSUE THREAD

**Issue by jonnyevans3210** [facebookresearch/detectron2#3517]
By altering the cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST config option before running inference, we get drastically different evaluation metric values. 

## Instructions To Reproduce the Issue:
Take the colab notebook as an example. Replace cell 12, `Train!`, with the following:
```
from detectron2.engine import DefaultTrainer
from detectron2.evaluation import COCOEvaluator, inference_on_dataset
from detectron2.data import build_detection_test_loader

results = []
cfg = get_cfg()
cfg.merge_from_file(model_zoo.get_config_file("COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml"))
cfg.DATASETS.TRAIN = ("balloon_train",)
cfg.DATASETS.TEST = ()
cfg.DATALOADER.NUM_WORKERS = 2
cfg.MODEL.WEIGHTS = model_zoo.get_checkpoint_url("COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml")  # Let training initialize from model zoo
cfg.SOLVER.IMS_PER_BATCH = 2
cfg.SOLVER.BASE_LR = 0.00025  # pick a good LR
cfg.SOLVER.MAX_ITER = 50    # 300 iterations seems good enough for this toy dataset; you will need to train longer for a practical dataset
cfg.SOLVER.STEPS = []        # do not decay learning rate
cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = 128   # faster, and good enough for this toy dataset (default: 512)
cfg.MODEL.ROI_HEADS.NUM_CLASSES = 1  # only has one class (ballon). (see https://detectron2.readthedocs.io/tutorials/datasets.html#update-the-config-for-new-datasets)
# NOTE: this config means the number of classes, but a few popular unofficial tutorials incorrect uses num_classes+1 here.

os.makedirs(cfg.OUTPUT_DIR, exist_ok=True)
trainer = DefaultTrainer(cfg) 
trainer.resume_or_load(resume=False)
trainer.train()
saved_cfg = cfg

results = []
for i in range(10):
  cfg = saved_cfg
  cfg.MODEL.WEIGHTS = os.path.join(cfg.OUTPUT_DIR, "model_final.pth")  # path to the model we just trained
  cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = i/10   # set a custom testing threshold
  predictor = DefaultPredictor(cfg)

  evaluator = COCOEvaluator("balloon_val", output_dir="./output")
  val_loader = build_detection_test_loader(cfg, "balloon_val")
  results.append((i/10, inference_on_dataset(predictor.model, val_loader, evaluator)['bbox']['AP50']))

results
```

The output is as follows (value of `results`):
```
[(0.0, 79.30198781745244),
 (0.1, 79.30198781745244),
 (0.2, 79.02937929464987),
 (0.3, 78.11410410973554),
 (0.4, 77.51446668120396),
 (0.5, 73.18508795959687),
 (0.6, 61.27812781278128),
 (0.7, 10.891089108910892),
 (0.8, nan),
 (0.9, nan)]
```

## Expected behavior:

I would expect the AP50 to be the same in each element of this list, regardless of the change in the threshold. The AP metrics should be calculated over the entire precision/recall curve by altering this threshold. By setting the threshold to anything other than 0, the evaluator has access to a smaller part of the precision/recall curve, and hence underestimates the performance of models. This means that the performance stated in the tutorial could be underestimated (as the default is 0.05), and I imagine many other projects.

I propose that at the least, this behaviour should be stated in the colab tutorial and README. Ideally, there would be a way to alter the config value to 0 when being inputted into the coco evaluator. Currently I do this manually in all my projects.

Note this issue was mentioned in a previous issue [here](https://github.com/facebookresearch/detectron2/issues/1484), but I think it was not fully realised or acted upon.

## Environment:

Detectron2 basic configuration. See colab tutorial example above for reproduction.

**Comment by ppwwyyxx** [comment]
> I propose that at the least, this behaviour should be stated in the colab tutorial and README

This is expected behavior and stated in the config documentation already, e.g. search for "ROI_HEADS.SCORE_THRESH_TEST" in https://detectron2.readthedocs.io/en/latest/modules/config.html#yaml-config-references
